### DISCUSSION TITLE
AutoGPT and newer AlpacaXGPT4 models

### DISCUSSION THREAD

**Discussion by anphex** [oobabooga/text-generation-webui#1243] [Ideas]
Hello!
Recently there are those 13b and 30b 4bit AlpacaXGPT4 models available on huggingface and they can be run on 12-24GB GPUs through the webui.

I wondered if with some tweaks or extensions it would be possible to combine this webui with AutoGPT to run locally.

I don't know if the context lenght max of 2048 (minus max new tokens) will become an issue and make it impractical.

**Comment by MithrilMan** [comment]
context lenght is a limiting factor, anyway I've started to experiment on this AUTO* stuff from scratch in C# for learn&research purpose and I'm going to use Vicuna as llm (I'm using the .pt version of the model here https://huggingface.co/TheBloke/vicuna-13B-1.1-GPTQ-4bit-128g )

I was surprised that I managed to get a prompt that at least returns me a JSON formatted output like I requested, here an example of the result

![image](https://user-images.githubusercontent.com/5107375/232255153-b7e69213-cc1b-4cef-895c-56d7ce829c90.png)

I started trying to replicate a similar AutoGPT JSON format, I just tried to focus on the first half of the format, thoughts, text, reasoning,etc.. are just garbage atm but the command part is on spot (actually I created commands from scratch without looking at AutoGPT way of doing it)
```
{
"command": "search",
"arguments": {
"query": "Einstein's most famous formula",
"max_results": 5
}
```
And this is enough for me to start to create an infrastructure around oobabooga API and start building from here

As a savvy full stack dev I'm tired to play with brittle interfaces and poor software architecture so I want to build my own stuff to play with and I'm sure this will be a long but funny journey!

UPDATE:
To people who's interested in following the development, I've created a repo here https://github.com/MithrilMan/AIdentities

**Reply by GitHub1712** [reply]
That would be cool, are you on a plugin or something?

**Reply by MithrilMan** [reply]
I'm actually creating a blazor server app with a plugin system in my spare time (not much atm)
I'm exploring at ways to achieve a plug&play plugin system loading assemblies on the fly straight from the UI, to have a kind of a dynamic system where people could code their own plugin
UI is made in MudBlazor to not have to waste too much time on creating components, I'm trying to abstract stuff and see how it comes togheter, it sounds like I'm already overengineering but I really feel the need to something more stable 

The choice about blazor server is because I envision this tool to be run on docker (I discourage everyone about just git clone on their system and run unknown code even more with these kind of repo that allow 3rd party plugin to mess with it without any security layer) and so running on localhost it fits well the case

**Reply by Felixqian4160** [reply]
It`s cool. I am very looking forward to it. I think use autogpt with vicuna13b will be very ineresting. even though I think if you could use two models one for autogpt one for thinking, let those two models work with each other, will be ineresting.

**Reply by MithrilMan** [reply]
Actually what I'd like to achieve in this plugin system is to abstract the connection to models as plugins too, so a plugin can literally be an implementation of a connector which is supposed to just connect to an API so if your system can then handle to load two different models of whatever kind you have to spin up for example two docker instance of oobobooga with two different loaded model and then query them both.

One of the first attempt I'll try to recreate is being able for instance to query both a text generation LLM (e.g. vicuna) and Automatic1111 stable diffusion API, I've a 3090TI but I'm not sure I'll be able to run them both, I suppose I'll find it out :)

Anyway, even if I'm focused on self hosted stuff on a single PC, playing with API allow you anyway to have multiple machines that interact in your local network, you will be free to use public API service too or multiple machines in local network of course.

I think I'll release a very basic version publicly as soon as I have a working pluggable model, then I would need to spin around the news to have people help implementing various connectors plugin to various API, etc..
Ambitious but nothing to lose because in any case I'll make use of it!

**Reply by GitHub1712** [reply]
Looking forward. I think a OPENAI API imitation would be great for text-generation-webui.

**Reply by MithrilMan** [reply]
that will be one of the first use case, yes, and one of the simplest I'd say

**Reply by Felixqian4160** [reply]
yes, it will be the first. and that is very cool, if you can run it

**Reply by SeattleUser0** [reply]
Imitating OpenAI's API would make it easy (or easier, I should say). People have been using Flask with it for quite some time already. Plenty of example code out there.

**Reply by Felixqian4160** [reply]
If one pc loading model with text-generation-webui could provide one api for another pc also loading model with text-generation-webui, that means we could use two pcs work with each other like autogpt with openai api？

**Reply by SeattleUser0** [reply]
> If one pc loading model with text-generation-webui could provide one api for another pc also loading model with text-generation-webui, that means we could use two pcs work with each other like autogpt with openai api？

That'd be one way to do it. Another way would be to use LangChain and Pinecone to hold information while text-generation-webui unloads and loads the appropriate models to retrieve solutions/answers. This would be quite beneficial as you could use one agent (Alpaca 30B) to read the goals and decide which models would work best on each step of the solution. The strategy would be similar to JARVIS, but being open source, the user could pick and choose custom models to add to the list. And if a good model didn't exist, you could make your own.

**Reply by Felixqian4160** [reply]
yes that is right，we could choose custom model， and then maybe will do something interesting

**Reply by SeattleUser0** [reply]
AutoGPT allows you to make "agents" and then assign them personalities. Ultimately, all agents are using the same source to determine the solution (GPT-4). It'd be like telling your alcoholic roommate (who has a really bad short-term memory) that he has a different occupation every morning. One day he wakes up and he's a firefighter, the next day a brain surgeon, and so on. He's inevitably going to suck at every single job you give him.

Custom models that are task specific are more accurate at their given task than a general model. MiniGPT-4 is one example of this. JARVIS is getting less attention than AutoGPT right now, but is ultimately the correct approach. I do not foresee a single model capable of doing everything coming along anytime in the near future, nor would I want to use it if it did.

As an example, if the goal was to create design plans for a building, you'd want multiple models -- likely one for each engineering discipline. You would not use the same model for drafting design as for architectural design. There's very little overlap on the Vin Diagram for these two activities.

**Reply by Felixqian4160** [reply]
Of course， SO everybody just training model with their own specialty, and then people can sell their own model to finish any jobs

**Reply by MithrilMan** [reply]
> If one pc loading model with text-generation-webui could provide one api for another pc also loading model with text-generation-webui, that means we could use two pcs work with each other like autogpt with openai api？

yes, at the end an endpoint could just be an IP that can point anywhere, on your local network on on the web

**Reply by MithrilMan** [reply]
> > If one pc loading model with text-generation-webui could provide one api for another pc also loading model with text-generation-webui, that means we could use two pcs work with each other like autogpt with openai api？
> 
> That'd be one way to do it. Another way would be to use LangChain and Pinecone to hold information while text-generation-webui unloads and loads the appropriate models to retrieve solutions/answers. This would be quite beneficial as you could use one agent (Alpaca 30B) to read the goals and decide which models would work best on each step of the solution. The strategy would be similar to JARVIS, but being open source, the user could pick and choose custom models to add to the list. And if a good model didn't exist, you could make your own.

yes having flexibility to chose model for your purpose it's a goal
Also pinecone could be exchanged by a self hosted solution like Redis

**Reply by MithrilMan** [reply]
I'm proceeding with my plugin architecture, here is an example of how I added a "chat plugin" (WIP)


https://user-images.githubusercontent.com/5107375/234880358-81f82a29-1fab-4f3f-815f-2be23535132b.mp4

**Reply by SeattleUser0** [reply]
Brilliant! I was modifying the methods in openai_.py in LoopGPT's release last night to talk to llama.py and thought to myself, "this would be awesome as a plug-in or extension in text-generation-webui." You're one step ahead as always!

**Reply by MithrilMan** [reply]
I've just "finished" (it will need probably more settings but the harder part is done) a section where you can create your character, I call it AIdentity, so you can have a repository of digital personalities that the user can then use to interact with whichever plugin he wants. here a video where I show the process of creation of such AIdentity

https://user-images.githubusercontent.com/5107375/235330993-11f4489c-5c1b-4ef8-ac5b-6583379e0941.mp4

Now my next step is to use this AIdentity on the chat session so that when you start a new session, you can chose who you want to talk with.
Then of course I need to interact with some proper API (I've to check current oobabooga API implementation) 
Then what I'd like to try is to implement a chat variation where you can create a conversation with multiple AIdentities and each one of them can reply to "itself" like if you were in a whatsapp group.
I've many idea to try out, maybe I should create a Discord server and try to drag in some people to help out and brainstorm about some nice idea to implement

**Reply by SeattleUser0** [reply]
Sign me up. I like where you’re going with this.

**Reply by MithrilMan** [reply]
for who's interested, I'm starting a discord server here the link:

https://discord.gg/5KbTuGQseB

Actually it's empty, just finished to setup channels, I still need to add bots ,etc.. (I'm not used to manage discord servers and actually I think the server has already what it's needed)

So please join to help shaping this new tool!

**Comment by sarutobiumon** [comment]
This would be truly revolutionary if it can AutoGPT can be used with local LLM models like the newer Alpaca or Vicuna ones.

**Reply by SeattleUser0** [reply]
There’s another project, LoopGPT, that looks like a better candidate for integration. But yes, the ability to use the GPT4-X-Alpaca model with an AutoGPT-like app would be amazing.

**Comment by Alampak** [comment]
Write current political situation in pakistan in 11May,2023
